[Enhancement] Support CSV header row for EXPORT and INSERT INTO FILES

tracymacding opened the pull request:
This PR adds support for including a header row in CSV file exports:

1. EXPORT statement: Add `with_header` property to include column names as the first row in exported CSV files. Example: EXPORT TABLE t TO "path" PROPERTIES ("with_header" = "true")

2. INSERT INTO FILES: Add `csv.include_header` property for the same functionality when using INSERT INTO FILES with CSV format. Example: INSERT INTO FILES ("format"="csv", "csv.include_header"="true", ...)

Changes:
- FE: Add property parsing in ExportStmt, ExportJob, ExportSink
- FE: Add csv.include_header support in TableFunctionTable
- BE: Add header writing logic in PlainTextBuilder for EXPORT
- BE: Add header writing logic in CSVFileWriter for INSERT INTO FILES
- Thrift: Add with_header field to TExportSink
- Thrift: Add csv_include_header field to TTableFunctionTable

## Why I'm doing:

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enables writing a header row (column names) in CSV outputs for both EXPORT and INSERT INTO FILES.
> 
> - FE: Parse `with_header` in `ExportStmt`; propagate via `ExportJob` and `ExportSink` (sets `column_names`/`with_header` in `TExportSink`).
> - FE: Parse `csv.include_header` in `TableFunctionTable`; include in `TTableFunctionTable`.
> - BE (EXPORT): `PlainTextBuilder` gains header support (escapes via `csv::escape_csv_field`) and is configured from `ExportSinkOperator`.
> - BE (INSERT INTO FILES): `CSVFileWriter`/Factory support `include_header` option, write header on first write/commit, escape special chars, and parse option.
> - Thrift: `TExportSink` adds `column_names` and `with_header`; `TTableFunctionTable` adds `csv_include_header`.
> - Tests: New/updated UTs for `PlainTextBuilder`, `CSVFileWriter`, and `ExportSinkOperator`; test list updated.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 20507ce67aa7d56ee9113b2729ace72a7489ea26. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

tracymacding committed:
[Enhancement] Support CSV header row for EXPORT and INSERT INTO FILES

This PR adds support for including a header row in CSV file exports:

1. EXPORT statement: Add `with_header` property to include column names
   as the first row in exported CSV files.
   Example: EXPORT TABLE t TO "path" PROPERTIES ("with_header" = "true")

2. INSERT INTO FILES: Add `csv.include_header` property for the same
   functionality when using INSERT INTO FILES with CSV format.
   Example: INSERT INTO FILES ("format"="csv", "csv.include_header"="true", ...)

Changes:
- FE: Add property parsing in ExportStmt, ExportJob, ExportSink
- FE: Add csv.include_header support in TableFunctionTable
- BE: Add header writing logic in PlainTextBuilder for EXPORT
- BE: Add header writing logic in CSVFileWriter for INSERT INTO FILES
- Thrift: Add with_header field to TExportSink
- Thrift: Add csv_include_header field to TTableFunctionTable

Signed-off-by: tracymacding <tracymacding@gmail.com>

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

kevincai commented on fe/fe-core/src/main/java/com/starrocks/load/ExportJob.java:
```diff
@@ -169,6 +169,8 @@ public class ExportJob implements Writable, GsonPostProcessable {
     @SerializedName("rd")
     private String rowDelimiter;
     private boolean includeQueryId;
+    @SerializedName("wh")
```
choose a different name, really confusing with `warehouse`

kevincai commented on be/src/exec/pipeline/sink/export_sink_operator.cpp:
```diff
@@ -109,10 +109,19 @@ Status ExportSinkIOBuffer::_open_file_writer() {
+
+    // Set header options if configured
+    if (_t_export_sink.__isset.with_header && _t_export_sink.with_header) {
+        builder_options.with_header = true;
```
what if with_header set to true but column_names is not set or is empty.

kevincai commented on be/src/formats/csv/csv_file_writer.cpp:
```diff
@@ -68,6 +70,20 @@ Status CSVFileWriter::init() {
     }
+
+    // Write header row if include_header is enabled
+    if (_writer_options->include_header && !_column_names.empty()) {
```
with this change, the `init()` will actually write something to the output stream, consider move this out of init phase, only do the header writing as necessary in `::write()` or `commit()` phase.

tracymacding committed:
[Enhancement] Address code review feedback for CSV header feature

1. Fix URI format in test: file:// -> file:///
2. Rename SerializedName from "wh" to "withHdr" for clarity
3. Add validation and warning log when with_header=true but column_names is empty
4. Simplify code style in plain_text_builder.cpp
5. Move header writing from init() to write() phase in csv_file_writer.cpp

Signed-off-by: tracymacding <tracymacding@gmail.com>

alvin-celerdata commented:
@cursor review

xiangguangyxg commented:
@Mergify backport branch-4.0

xiangguangyxg commented:
@Mergify backport branch-3.5
